### PR TITLE
feat(Liquidation: SignerTimeout): finish SignerTimeout logics

### PR DESCRIPTION
**Pull request by cosinlink**

### CODE REVIEW

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 1]
```diff
@@ -1,7 +1,73 @@
...
-use crate::utils::types::Error;
+use crate::utils::types::{
+    Error, XChainKind
+};
 use core::result::Result;
+use ckb_std::{
+    ckb_constants::Source,
+    high_level::{load_cell_capacity}
+};
+use ckb_std::high_level::load_input_since;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:10]: only one `use`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 2]
```diff
@@ -1,7 +1,73 @@
...
 
-pub fn verify(_toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
+pub fn verify(toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
+    if toCKB_data_tuple.0.is_none() {
+        return Err(Error::InputDataInvalid)
+    }
+    if toCKB_data_tuple.1.is_none() {
+        return Err(Error::OutputDataInvalid)
+    }
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:18]: not need to judge, those conditions are impossible

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 3]
```diff
@@ -1,7 +1,73 @@
...
+
+    // `get_toCKB_data_tuple` has checked cell nums == 1 in both inputs and outputs
+    // Todo: check since n4
+    let _since = load_input_since(0, Source::GroupInput);
+
+    // check if capacity and data[2..7] are not changed
+    let cap_input = load_cell_capacity(0, Source::GroupInput)
+        .map_err(|_| Error::CapacityInvalid)?;
+    let cap_output = load_cell_capacity(0, Source::GroupOutput)
+        .map_err(|_| Error::CapacityInvalid)?;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:31]: `let cap_input = load_cell_capacity(0, Source::GroupInput)?;`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 4]
```diff
@@ -1,7 +1,73 @@
...
+    if toCKB_data_tuple.1.is_none() {
+        return Err(Error::OutputDataInvalid)
+    }
+
+    let input_data = toCKB_data_tuple.0.as_ref().unwrap();
+    let output_data = toCKB_data_tuple.1.as_ref().unwrap();
+
+    // `get_toCKB_data_tuple` has checked cell nums == 1 in both inputs and outputs
+    // Todo: check since n4
+    let _since = load_input_since(0, Source::GroupInput);
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:25]: use another `fn verify_since`

File: contracts/toCKB-typescript/src/utils/types/error.rs [hunk 5]
```diff
@@ -12,6 +12,19 @@ pub enum Error {
...
+    ToCKBStatusMismatch,
+    CapacityInvalid,
+    KindInvalid,
+    BTCLotSizeInvalid,
+    ETHLotSizeInvalid,
+    UserLockInvalid,
+    XLockAddressInvalid,
+    SignerLockInvalid,
+    XUnlockAddressInvalid,
+    RedeemerLockInvalid
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/utils/types/error.rs:27]: you only need two errors: `InvariantDataMutated`, `InputSinceInvalid`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 6]
```diff
@@ -1,7 +1,73 @@
...
+pub fn verify(toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
+    if toCKB_data_tuple.0.is_none() {
+        return Err(Error::InputDataInvalid)
+    }
+    if toCKB_data_tuple.1.is_none() {
+        return Err(Error::OutputDataInvalid)
+    }
+
+    let input_data = toCKB_data_tuple.0.as_ref().unwrap();
+    let output_data = toCKB_data_tuple.1.as_ref().unwrap();
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:21]: `toCKB_data_tuple.0.as_ref().expect("inputs contain toCKB cell")`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 7]
```diff
@@ -1,7 +1,55 @@
...
+        || input_data.user_lockscript_hash.as_ref() != output_data.user_lockscript_hash.as_ref()
+        || input_data.x_lock_address.as_ref() != output_data.x_lock_address.as_ref()
+        || input_data.signer_lockscript_hash.as_ref() != output_data.signer_lockscript_hash.as_ref()
+        || input_data.x_unlock_address.as_ref() != output_data.x_unlock_address.as_ref()
+        || input_data.redeemer_lockscript_hash.as_ref() != output_data.redeemer_lockscript_hash.as_ref()
+    {
+        return Err(Error::InvariantDataMutated);
+    }
+
+    match input_data.kind {
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:37]: Should be fixed after @mkxbl implement a `get_lot_size` function.

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 8]
```diff
@@ -1,7 +1,55 @@
...
+        }
+    };
 
-pub fn verify(_toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
     Ok(())
 }
+
+fn verify_since(_since: u64) -> Result<(), Error> {
+    Ok(())
+}
```
Comment (huwenchao) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:55]:
```suggestion
}

```

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs [hunk 9]
```diff
@@ -1 +1,208 @@
...
+use ckb_tool::{ckb_error::assert_error_eq, ckb_script::ScriptError};
+use molecule::prelude::*;
+use int_enum::IntEnum;
 
+
+const MAX_CYCLES: u64 = 10_000_000;
+const PLEDGE_INVALID: i8 = 8;
+const LOT_SIZE_INVALID: i8 = 7;
+const TX_INVALID: i8 = 6;
+const ENCODING: i8 = 4;
```
Comment (huwenchao) [inline tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:19]: The error code and `ToCKBStatus` are duplicated here.
When the contract changes, it is hard to change it accordingly.
Consider move the definitions into a separated file, and use a softlink in the tests or make a rule in Makefile to copy it to tests.

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 8 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:55]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs (hunk 9 above)
Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:19]: This change may affect many people's rebase.Maybe it is better to open another PR and unify the standard code fmt? @mkxbl

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 10]
```diff
@@ -1,7 +1,52 @@
...
+
+fn verify_lot_size(toCKB_data: &ToCKBCellDataView) -> Result<(), Error> {
+    let xchain_kind = get_xchain_kind()?;
+    match xchain_kind {
+        XChainKind::Btc if toCKB_data.get_btc_lot_size().is_ok() => Ok(()),
+        XChainKind::Eth if toCKB_data.get_eth_lot_size().is_ok() => Ok(()),
+        _ => Err(Error::LotSizeInvalid),
+    }
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:51]: This is incorrect. What you need to do is verify `lot_size` is unchanged, so use `fn get_raw_lot_size` to deal with it

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 11]
```diff
@@ -1,7 +1,52 @@
...
+};
 
-pub fn verify(_toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
+pub fn verify(toCKB_data_tuple: &ToCKBCellDataTuple) -> Result<(), Error> {
+    let input_data = toCKB_data_tuple.0.as_ref().expect("inputs should contain toCKB cell");
+    let output_data = toCKB_data_tuple.1.as_ref().expect("outputs should contain toCKB cell");
+
+    // `get_toCKB_data_tuple` has checked cell nums == 1 in both inputs and outputs
+    // Todo: check since n4
+    let since = load_input_since(0, Source::GroupInput).expect("since should exist");
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:18]: move line 18 to the function `fn verify_since`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 12]
```diff
@@ -1,7 +1,52 @@
...
+    if input_data.user_lockscript.as_ref() != output_data.user_lockscript.as_ref()
+        || input_data.x_lock_address.as_ref() != output_data.x_lock_address.as_ref()
+        || input_data.signer_lockscript.as_ref() != output_data.signer_lockscript.as_ref()
+        || input_data.x_unlock_address.as_ref() != output_data.x_unlock_address.as_ref()
+        || input_data.redeemer_lockscript.as_ref() != output_data.redeemer_lockscript.as_ref()
+    {
+        return Err(Error::InvariantDataMutated);
+    }
+
+    verify_lot_size(input_data)?;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:37]: not need another function here

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 13]
```diff
@@ -1,7 +1,52 @@
...
+    // Todo: check since n4
+    let since = load_input_since(0, Source::GroupInput).expect("since should exist");
+    verify_since(since)?;
+
+    // check if capacity and data[2..7] are not changed
+    let cap_input = load_cell_capacity(0, Source::GroupInput).expect("get input capacity");
+    let cap_output = load_cell_capacity(0, Source::GroupOutput).expect("get output capacity");
+    if cap_input != cap_output {
+        return Err(Error::CapacityInvalid);
+    }
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:26]: maybe use another `fn verify_capacity`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 14]
```diff
@@ -1,7 +1,52 @@
...
+
+    if input_data.user_lockscript.as_ref() != output_data.user_lockscript.as_ref()
+        || input_data.x_lock_address.as_ref() != output_data.x_lock_address.as_ref()
+        || input_data.signer_lockscript.as_ref() != output_data.signer_lockscript.as_ref()
+        || input_data.x_unlock_address.as_ref() != output_data.x_unlock_address.as_ref()
+        || input_data.redeemer_lockscript.as_ref() != output_data.redeemer_lockscript.as_ref()
+    {
+        return Err(Error::InvariantDataMutated);
+    }
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:35]: maybe use another `fn verify_data`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 10 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 11 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 13 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:35]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 14 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:47]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 12 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs (hunk 9 above)
Comment (mkxbl) [reply tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:19]: `Error` `ToCKBStatus` `LotSize` `XChainKind`  are added to utils/types

Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:19]: ok

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 15]
```diff
@@ -1,7 +1,63 @@
...
+            if input_data.get_btc_lot_size()? != output_data.get_btc_lot_size()? {
+                return Err(Error::InvariantDataMutated);
+            }
+        }
+        XChainKind::Eth => {
+            if input_data.get_eth_lot_size()? != output_data.get_eth_lot_size()? {
+                return Err(Error::InvariantDataMutated);
+            }
+        }
+    };
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:60]: use `fn get_raw_lot_size`: 
``` rust
input_data.get_raw_lot_size() == output_data.get_raw_lot_size()
```

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 16]
```diff
@@ -1,7 +1,63 @@
...
+    let output_data = toCKB_data_tuple.1.as_ref().expect("outputs should contain toCKB cell");
+
+    verify_since()?;
+    verify_capacity()?;
+    verify_data(input_data, output_data)?;
+    Ok(())
+}
+
+fn verify_since() -> Result<(), Error> {
+    // Todo: check since n4
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:23]: fix TODO

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs [hunk 17]
```diff
@@ -1 +1,182 @@
...
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:19]: remove error const

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs [hunk 18]
```diff
@@ -1 +1,182 @@
...
+pub enum ToCKBStatus {
+    Initial = 1,
+    Bonded = 2,
+    Warranty = 3,
+    Redeeming = 4,
+    SignerTimeout = 5,
+    Undercollateral = 6,
+    FaultyWhenWarranty = 7,
+    FaultyWhenRedeeming = 8,
+}
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:33]: remove it, import from utils/types instead

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs [hunk 19]
```diff
@@ -1 +1,182 @@
...
+        .status(Byte::new(ToCKBStatus::SignerTimeout as u8))
+        .lot_size(Byte::new(2u8))
+        .build();
+
+    let (context, tx) =
+        build_test_context(1,input_toCKB_data.as_bytes(), output_toCKB_data.as_bytes());
+
+
+    let err = context.verify_tx(&tx, MAX_CYCLES).unwrap_err();
+    assert_error_eq!(err, ScriptError::ValidationFailure(INVARIANT_DATA_MUTATED));
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:74]: `assert_error_eq!(err, ScriptError::ValidationFailure(Error::InvariantDataMutated));`

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 15 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:71]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 16 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs (hunk 18 above)
Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs (hunk 19 above)
Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 20]
```diff
@@ -1,7 +1,60 @@
...
+        || since & METRIC_TYPE_FLAG_MASK != SINCE_TYPE_TIMESTAMP {   // check if it is timestamp value
+        return Err(Error::InputSinceInvalid)
+    }
+
+    let time = since & VALUE_MASK;
+    if time != N4 {
+        return Err(Error::InputSinceInvalid)
+    }
+
+    Ok(())
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:36]: please ref: https://github.com/nervosnetwork/axon-internal/issues/66#issuecomment-680786173

File: contracts/toCKB-typescript/src/utils/config.rs [hunk 21]
```diff
@@ -1 +1,11 @@
 pub const PLEDGE: u64 = 10000;
+
+// since
+pub const LOCK_TYPE_FLAG: u64 = 1 << 63;
+pub const METRIC_TYPE_FLAG_MASK: u64 = 0x6000_0000_0000_0000;
+pub const VALUE_MASK: u64 = 0x00ff_ffff_ffff_ffff;
+pub const REMAIN_FLAGS_BITS: u64 = 0x1f00_0000_0000_0000;
+pub const SINCE_TYPE_TIMESTAMP: u64 = 0x4000_0000_0000_0000;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/utils/config.rs:8]: please ref: https://github.com/nervosnetwork/axon-internal/issues/66#issuecomment-680786173

Comment (cosinlink) [reply contracts/toCKB-typescript/src/utils/config.rs]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs (hunk 20 above)
Comment (cosinlink) [reply contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs]: fixed

File: tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs [hunk 22]
```diff
@@ -1 +1,171 @@
...
+        CellOutput::new_builder()
+            .capacity(capacity.pack())
+            .lock(always_success_lockscript.clone())
+            .type_(Some(toCKB_typescript.clone()).pack())
+            .build(),
+        input_toCKB_data,
+    );
+    let input = CellInput::new_builder()
+        .previous_output(input_out_point)
+        .since(SINCE_SIGNER_TIMEOUT.pack())
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs:149]: miss `fn test_wrong_since`

Comment (cosinlink) [reply tests/src/toCKB_typescript/tests/liquidation_signertimeout.rs]: fixed

File: contracts/toCKB-typescript/src/utils/config.rs [hunk 23]
```diff
@@ -1 +1,7 @@
 pub const PLEDGE: u64 = 10000;
+
+// since
+pub const LOCK_TYPE_FLAG: u64 = 1 << 63;
+pub const SINCE_TYPE_TIMESTAMP: u64 = 0x4000_0000_0000_0000;
+pub const N4: u64 = 3 * 24 * 3600; // timeout
```
[the remark this answers is not in the document]
Comment (cosinlink) [reply contracts/toCKB-typescript/src/utils/config.rs:6]: fixed

File: contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs [hunk 24]
```diff
@@ -1,7 +1,61 @@
...
+    verify_since()?;
+    verify_capacity()?;
+    verify_data(input_data, output_data)?;
+    Ok(())
+}
+
+fn verify_since() -> Result<(), Error> {
+    let since = load_input_since(0, Source::GroupInput).map_err(|_| Error::InputSinceInvalid)?;
+    if since != SINCE_SIGNER_TIMEOUT {
+        // check if it is timestamp value
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/switch/liquidation_signertimeout/mod.rs:31]: remove useless comment

File: contracts/toCKB-typescript/src/utils/config.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (mkxbl) [reply contracts/toCKB-typescript/src/utils/config.rs:6]: > fixed

It seems not. 

And if you prefer timestamp, you should note that the unit is millisecond

File: contracts/toCKB-typescript/src/utils/config.rs [hunk 25]
```diff
@@ -1 +1,8 @@
 pub const PLEDGE: u64 = 10000;
+
+// since
+pub const LOCK_TYPE_FLAG: u64 = 1 << 63;
+pub const SINCE_TYPE_TIMESTAMP: u64 = 0x4000_0000_0000_0000;
+
+// 3 * 24 * 3600 means 3 days, the unit is millisecond
+pub const SINCE_SIGNER_TIMEOUT: u64 = LOCK_TYPE_FLAG | SINCE_TYPE_TIMESTAMP | 3 * 24 * 3600;
```
Comment (mkxbl) [inline contracts/toCKB-typescript/src/utils/config.rs:8]: I'm falling apart  😂

```suggestion
pub const SINCE_SIGNER_TIMEOUT: u64 = LOCK_TYPE_FLAG | SINCE_TYPE_TIMESTAMP | 1000 * 24 * 3600;
```

File: tests/src/toCKB_typescript/utils/config.rs [hunk 26]
```diff
@@ -0,0 +1,8 @@
+pub const PLEDGE: u64 = 10000;
+
+// since
+pub const LOCK_TYPE_FLAG: u64 = 1 << 63;
+pub const SINCE_TYPE_TIMESTAMP: u64 = 0x4000_0000_0000_0000;
+
+// 3 * 24 * 3600 means 3 days, the unit is millisecond
+pub const SINCE_SIGNER_TIMEOUT: u64 = LOCK_TYPE_FLAG | SINCE_TYPE_TIMESTAMP | 3 * 24 * 3600;
```
Comment (mkxbl) [inline tests/src/toCKB_typescript/utils/config.rs:8]: I'm falling apart  😂

```suggestion
pub const SINCE_SIGNER_TIMEOUT: u64 = LOCK_TYPE_FLAG | SINCE_TYPE_TIMESTAMP | 1000 * 24 * 3600;
```

**Pull request merged**